준비상태에 있는 오리들만 보이게 수정

sbyeol3 opened the pull request:
## 💁 설명

### DuckObject & PlayerManager 수정
- DuckObject에서는 바로 이벤트 등록을 해주고 있는데 보이지 않을 때 계속 리스너를 해줄 필요는 없을 것 같아서?! 일단 분리를 해두었습니다.
- visibility에 따라서 이벤트 등록 / 제거를 합니다.
- `ready player` 이벤트의 콜백으로 true일 때는 해당 오리를 보이게 false일 때는 보이지 않게 했습니다~

### 기존 플레이어의 프로필
- 이전에는 닉네임과 컬러만 받아왔다면 ready도 같이 넘겨줘서 만약 이전에 레디한 상태라면 처음 들어온 유저가 볼 수 있게 수정

### 일부 리팩토링
- game.js에 들어있던 소켓 이벤트 콜백함수들을 waitingroom 으로 이동했습니다.
- 기존 레디버튼 관리를 dataset으로 했는데 이거를 player로 ready상태 관리를 빼면서 player를 이용해서 했습니다~

![화면 기록 2020-12-10 오후 3 35 35 mov](https://user-images.githubusercontent.com/43198553/101730504-7a6c8000-3afd-11eb-8490-965c708aec5a.gif)

## 📑 체크리스트

> 구현한 목록 체크리스트
- [ ] 준비오리만 볼 수 있다.

## 🚧 주의 사항

> PR을 읽을 때 살펴볼 사항

- 주의 사항 1
- 주의 사항 2

sbyeol3 committed:
Feat : Player의 커서 오리 컨트롤하는 함수 추가 (#174)

- 플레이어의 준비 상태 관리하는 함수 추가
- set한 플레이어가 본인일경우 event 리스너 컨트롤
- 아니면 그냥 visibility만 수정

sbyeol3 committed:
Refactort : game의 소켓 이벤트를 waitingRoom으로 이동

- 대기실에서만 발생하는 것들은 대기실 소켓파일로 이동

sbyeol3 committed:
Feat : 마우스이벤트 리스너 수정 (#174)

- 보이지 않을 때는 마우스무브에 대한 이벤트 리스너를 해제해도 되므로 수정

Front-line-dev commented on src/frontend/engine/DuckCursorObject.js:
```diff
@@ -31,23 +31,32 @@ class DuckCursorObject extends DuckObejct {
-        }
-      }, TIME.DUCK_THROTTLE);
-    });
+  removeMouseMoveEvent() {
+    $id('root').removeEventListener('mousemove', (event) =>
+      this.makeFollowMouse(event),
+    );
+  }
```
remove를 구현해주셨군요! 감사합니다

ramram1048 commented on src/frontend/engine/DuckCursorObject.js:
```diff
@@ -31,23 +31,32 @@ class DuckCursorObject extends DuckObejct {
-      this.moveToMouse(calcPosition(event));
-      setTimeout(() => {
-        this.throttling = false;
-        if (this.lastPosition) {
-          this.moveToMouse(this.lastPosition);
-          this.lastPosition = null;
-        }
-      }, TIME.DUCK_THROTTLE);
-    });
+  removeMouseMoveEvent() {
+    $id('root').removeEventListener('mousemove', (event) =>
+      this.makeFollowMouse(event),
+    );
+  }
```
이거 브라우저 개발자도구로 EventListener 확인했을때 잘 없어지는거 확인되나요 ??
35～36라인의 `(event) => this.makeFollowMouse(event)`와 41～42라인의 `(event) => this.makeFollowMouse(event)`가 다른 함수로 인식되면 삭제되지 않을 것 같아서요. 
혹시 잘 없어진다면 괜찮지만 안없어지는게 맞다면 그냥 `this.makeFollowMouse`로 써주시면 될것같아요.

sbyeol3 replied:
ㅇㅏ 왜 저렇게 했었더라... 맞아요 저거 절케해야하는디... 머엿드라... this가 바인딩이 안됐던거같다 확인해볼게요

ramram1048 reviewed (commented):
Game.js가 가벼워지는 첫 걸음이네요 👍

jinhyukoo reviewed (approved):
좋습니당~~~ 기획한게 하나씩 만들어지고 있군요~~~

Front-line-dev commented on src/frontend/utils/Player.js:
```diff
@@ -29,13 +26,30 @@ const Player = class {
     this.duck.delete();
   }
+
+  setVisibility(value = false) {
+    const displayStyle = value ? 'block' : 'none';
+    this.duck.instance.style.display = displayStyle;
+
+    if (this.isCurrentPlayer) this.setMouseEvent(value);
+  }
+
+  setMouseEvent(value = true) {
+    if (value) this.duck.addMouseMoveEvent();
+    else this.duck.removeMouseMoveEvent();
+  }
```
duck에 관한 함수는 duck에다 만드는게 나을거 같아요. PlayerManager -> Player -> duck 이렇게 접근을 하는데 duck을 조정하려고 Player에 만들면 PlayerManager에도 만들게 되고... 그렇지 않을까요? 물론 클래스 내부에 접근 할 때는 메소드를 써서 접근하는게 정석이지만 그러면 차라리 getDuck()을 만드는게 나을거 같아요. 개별 플레이어의 duck에 접근하기가 어려워서 만드신거라면 나중에 감당이 안 될거 같네요

Front-line-dev reviewed (commented):
수고하셨습니다!

sbyeol3 committed:
Fix : removeEventHandler 안되는 문제 해결

- bind를 해주어야 하는데 add함수에서 넣어주면 안되고 미리 bind한 함수를 선언한다

sbyeol3 committed:
Feat : DuckObject 수정 및 기존 플레이어 ready 설정 (#174)

- 기존 플레이어 레디상태도 받아와서 ready로 초기 visibility 변경
- player에 있던 메소드를 DuckCursor 클래스로 옮김

sbyeol3 committed:
Feat : 서버 userProfile에서 ready도 넘겨줌

- 기존 플레이어 레디 상태도 같이 보냄
